### PR TITLE
Wasmer interface refactor, part 1: the exports

### PR DESCRIPTION
**Pull request by andrei-marinica**
Wasmer export pointers removed everywhere other than the wasmer package.

The executor interface no longer contains any wasmer export pointers. They are an implementation detail.

Part 2 will see further interface reorg, I didn't want to clutter this.

Also note that `elrondapimeta` is getting renamed to `executorinterface` in part 2, so keep this in mind when reading. The name made sense when it only contained the imports, but it now goes both ways.

### CODE REVIEW

File: arwen/contexts/output.go [hunk 1]
```diff
@@ -5,12 +5,13 @@ import (
...
 	"math/big"
 
-	"github.com/ElrondNetwork/wasm-vm/arwen"
 	"github.com/ElrondNetwork/elrond-go-core/core"
 	"github.com/ElrondNetwork/elrond-go-core/core/check"
 	"github.com/ElrondNetwork/elrond-go-core/data/vm"
 	logger "github.com/ElrondNetwork/elrond-go-logger"
 	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+	"github.com/ElrondNetwork/wasm-vm/arwen"
+	"github.com/ElrondNetwork/wasm-vm/arwen/elrondapimeta"
```
Comment (sasurobert) [inline arwen/contexts/output.go:14]: make it only apimeta

File: arwen/contexts/runtime.go [hunk 2]
```diff
@@ -736,40 +731,26 @@ func (context *runtimeContext) CountSameContractInstancesOnStack(address []byte)
...
-
-	return nil
+	return "", elrondapimeta.ErrFuncNotFound
 }
 
-// HasCallbackMethod returns true if the current wasmer instance exports has a callback method.
-func (context *runtimeContext) HasCallbackMethod() bool {
-	_, ok := context.instance.GetExports()[arwen.CallbackFunctionName]
-	return ok
+func (context *runtimeContext) CallSCFunction(functionName string) error {
```
Comment (sasurobert) [inline arwen/contexts/runtime.go:752]: add comment to public function

File: arwen/contexts/runtime.go [hunk 3]
```diff
@@ -898,15 +879,14 @@ func (context *runtimeContext) ValidateCallbackName(callbackName string) error {
...
 	}
 	if !context.HasFunction(callbackName) {
-		return arwen.ErrFuncNotFound
+		return elrondapimeta.ErrFuncNotFound
 	}
 
 	return nil
 }
 
 func (context *runtimeContext) HasFunction(functionName string) bool {
```
Comment (sasurobert) [inline arwen/contexts/runtime.go:888]: add comment to public function

File: mock/context/instanceMock.go [hunk 4]
```diff
@@ -95,23 +96,30 @@ func (instance *InstanceMock) Cache() ([]byte, error) {
 func (instance *InstanceMock) Clean() {
 }
 
-// GetExports mocked method
-func (instance *InstanceMock) GetExports() wasmer.ExportsMap {
-	return instance.Exports
+func (instance *InstanceMock) CallFunction(functionName string) error {
```
Comment (sasurobert) [inline mock/context/instanceMock.go:99]: add comments.

File: mock/context/instanceMock.go [hunk 5]
```diff
@@ -95,23 +96,30 @@ func (instance *InstanceMock) Cache() ([]byte, error) {
...
+		_, err := function()
+		return err
+	}
+
+	return elrondapimeta.ErrFuncNotFound
 }
 
-// GetSignature mocked method
-func (instance *InstanceMock) GetSignature(functionName string) (*wasmer.ExportedFunctionSignature, bool) {
+func (instance *InstanceMock) HasFunction(functionName string) bool {
```
Comment (sasurobert) [inline mock/context/instanceMock.go:108]: add comments

File: mock/context/instanceMock.go [hunk 6]
```diff
@@ -95,23 +96,30 @@ func (instance *InstanceMock) Cache() ([]byte, error) {
...
 	_, ok := instance.Exports[functionName]
+	return ok
+}
 
-	if !ok {
-		return nil, false
+func (instance *InstanceMock) GetFunctionNames() []string {
```
Comment (sasurobert) [inline mock/context/instanceMock.go:113]: add comments

File: mock/context/instanceMock.go [hunk 7]
```diff
@@ -95,23 +96,30 @@ func (instance *InstanceMock) Cache() ([]byte, error) {
...
+		functionNames = append(functionNames, functionName)
 	}
+	return functionNames
+}
 
-	return &wasmer.ExportedFunctionSignature{
-		InputArity:  0,
-		OutputArity: 0,
-	}, true
+func (instance *InstanceMock) ValidateVoidFunction(functionName string) error {
```
Comment (sasurobert) [inline mock/context/instanceMock.go:121]: add comments

File: mock/context/runtimeContextMock.go [hunk 8]
```diff
@@ -236,26 +236,20 @@ func (r *RuntimeContextMock) GetInstance() wasmer.InstanceHandler {
...
+		return "", r.Err
 	}
-	return nil, nil
+	return "", nil
 }
 
-// GetInitFunction mocked method
-func (r *RuntimeContextMock) GetInitFunction() wasmer.ExportedFunctionCallback {
-	return nil
+func (r *RuntimeContextMock) CallSCFunction(functionName string) error {
```
Comment (sasurobert) [inline mock/context/runtimeContextMock.go:251]: add comment

File: mock/context/runtimeContextWrapper.go [hunk 9]
```diff
@@ -487,19 +481,12 @@ func (contextWrapper *RuntimeContextWrapper) GetInstance() wasmer.InstanceHandle
...
 
-// GetInstanceExports calls corresponding xxxFunc function, that by default in turn calls the original method of the wrapped RuntimeContext
-func (contextWrapper *RuntimeContextWrapper) GetInstanceExports() wasmer.ExportsMap {
-	return contextWrapper.GetInstanceExportsFunc()
-}
-
-// GetInitFunction calls corresponding xxxFunc function, that by default in turn calls the original method of the wrapped RuntimeContext
-func (contextWrapper *RuntimeContextWrapper) GetInitFunction() wasmer.ExportedFunctionCallback {
-	return contextWrapper.GetInitFunctionFunc()
+func (contextWrapper *RuntimeContextWrapper) FunctionNameChecked() (string, error) {
```
Comment (sasurobert) [inline mock/context/runtimeContextWrapper.go:484]: comment comments everywhere

File: wasmer/instance.go [hunk 10]
```diff
@@ -313,15 +315,30 @@ func (instance *Instance) IsFunctionImported(name string) bool {
 	return cWasmerInstanceIsFunctionImported(instance.instance, name)
 }
 
-// GetExports returns the exports map for the current instance
-func (instance *Instance) GetExports() ExportsMap {
-	return instance.Exports
+func (instance *Instance) CallFunction(functionName string) error {
```
Comment (sasurobert) [inline wasmer/instance.go:318]: comment

File: wasmer/instance.go [hunk 11]
```diff
@@ -313,15 +315,30 @@ func (instance *Instance) IsFunctionImported(name string) bool {
...
+	if function, ok := instance.Exports[functionName]; ok {
+		_, err := function()
+		return err
+	}
+
+	return elrondapimeta.ErrFuncNotFound
+}
+
+func (instance *Instance) HasFunction(functionName string) bool {
```
Comment (sasurobert) [inline wasmer/instance.go:327]: comment

File: wasmer/instance.go [hunk 12]
```diff
@@ -313,15 +315,30 @@ func (instance *Instance) IsFunctionImported(name string) bool {
...
+	_, ok := instance.Exports[functionName]
+	return ok
+}
+
+func (instance *Instance) GetFunctionNames() []string {
```
Comment (sasurobert) [inline wasmer/instance.go:332]: comment

File: wasmer/instance.go [hunk 13]
```diff
@@ -313,15 +315,30 @@ func (instance *Instance) IsFunctionImported(name string) bool {
...
+		functionNames = append(functionNames, functionName)
+	}
+	return functionNames
 }
 
-// GetSignature returns the signature for the given functionName
-func (instance *Instance) GetSignature(functionName string) (*ExportedFunctionSignature, bool) {
-	signature, ok := instance.Signatures[functionName]
-	return signature, ok
+func (instance *Instance) ValidateVoidFunction(functionName string) error {
```
Comment (sasurobert) [inline wasmer/instance.go:340]: comment

File: wasmer/interface.go [hunk 14]
```diff
@@ -29,3 +31,7 @@ type MemoryHandler interface {
 	Destroy()
 	IsInterfaceNil() bool
 }
+
+type ValidatorContext interface {
```
Comment (sasurobert) [inline wasmer/interface.go:35]: add comment for interface

Comment (sasurobert) [reply wasmer/interface.go:35]: IsInterfaceNil() bool

File: arwen/contexts/output.go (hunk 1 above)
Comment (andrei-marinica) [reply arwen/contexts/output.go:14]: it's renamed to `executorinterface` in the next PR

File: wasmer/interface.go (hunk 14 above)
Comment (andrei-marinica) [reply wasmer/interface.go]: Turns out it was unused, deleted altogether.

File: arwen/contexts/runtime.go [hunk 15]
```diff
@@ -736,40 +731,27 @@ func (context *runtimeContext) CountSameContractInstancesOnStack(address []byte)
...
 
-// GetFunctionToCall returns the callable contract method to be executed, as exported by the Wasmer instance.
-func (context *runtimeContext) GetFunctionToCall() (wasmer.ExportedFunctionCallback, error) {
-	exports := context.instance.GetExports()
-	logRuntime.Trace("get function to call", "function", context.callFunction)
-	if function, ok := exports[context.callFunction]; ok {
-		return function, nil
+// FunctionNameChecked returns the function name, after checking that it exists in the contract.
+func (context *runtimeContext) FunctionNameChecked() (string, error) {
+	functionName := context.Function()
```
Comment (camilbancioiu) [inline arwen/contexts/runtime.go:736]: Consider renaming `context.Function()` to `context.FunctionName()`.

File: test/contracts/signatures/signatures.c [hunk 16]
```diff
@@ -1,26 +1,19 @@
 #include "../elrond/context.h"
```
Comment (camilbancioiu) [inline test/contracts/signatures/signatures.c:1]: `context.h` can be removed. The type `byte` is defined in `../elrond/types.h`.

**Review by camilbancioiu — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**